File: sandeep_reports/sandeep_reports/report/store_shift_report/store_shift_report.js
// Store Shift Report — pick a shift date + shift, see Inward/Outward store movements, print or export.
frappe.query_reports["Store Shift Report"] = {
	filters: [
		{
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
			default: frappe.defaults.get_user_default("Company"),
			reqd: 1,
		},
		{
			fieldname: "shift_date",
			label: __("Shift Date"),
			fieldtype: "Date",
			default: frappe.datetime.get_today(),
			reqd: 1,
		},
		{
			fieldname: "shift",
			label: __("Shift"),
			fieldtype: "Select",
			options: [
				{ value: "Full Day", label: __("Full Day (00:00–24:00)") },
				{ value: "A", label: __("Shift A (06:00–14:00)") },
				{ value: "B", label: __("Shift B (14:00–22:00)") },
				{ value: "C", label: __("Shift C (22:00–06:00 next day)") },
			],
			default: "Full Day",
			reqd: 1,
		},
		{
			fieldname: "movement",
			label: __("Movement"),
			fieldtype: "Select",
			options: ["All", "Inward", "Outward"],
			default: "All",
		},
		{
			fieldname: "warehouse",
			label: __("Warehouse / Store"),
			fieldtype: "Link",
			options: "Warehouse",
			get_query: function () {
				var c = frappe.query_report.get_filter_value("company");
				return { filters: c ? { company: c } : {} };
			},
		},
		{
			fieldname: "item_group",
			label: __("Item Group"),
			fieldtype: "Link",
			options: "Item Group",
		},
		{
			fieldname: "item_code",
			label: __("Item"),
			fieldtype: "Link",
			options: "Item",
		},
	],

	formatter: function (value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		if (!data) return value;

		// Section header rows: bold banner across the Voucher Type cell.
		if (data.row_type === "section") {
			if (column.fieldname === "voucher_type") {
				return "<span style='font-weight:700;font-size:13px;color:#1a3c5e'>" + (data.voucher_type || "") + "</span>";
			}
			return "";
		}
		// Subtotal rows
		if (data.row_type === "subtotal") {
			if (column.fieldname === "party") {
				return "<span style='font-weight:700;color:#1a3c5e'>" + (data.party || "") + "</span>";
			}
			if (column.fieldname === "in_qty" || column.fieldname === "out_qty") {
				return "<span style='font-weight:700'>" + value + "</span>";
			}
			if (column.fieldname === "balance_qty" || column.fieldname === "uom" ||
				column.fieldname === "voucher_type" || column.fieldname === "voucher_no" ||
				column.fieldname === "item_code" || column.fieldname === "item_name" ||
				column.fieldname === "warehouse" || column.fieldname === "posting_time") {
				return "";
			}
			return value;
		}
		// Data rows: colour the qty columns
		if (column.fieldname === "in_qty" && (parseFloat(data.in_qty) || 0) > 0) {
			value = "<span style='color:#2980b9'>" + value + "</span>";
		}
		if (column.fieldname === "out_qty" && (parseFloat(data.out_qty) || 0) > 0) {
			value = "<span style='color:#e67e22'>" + value + "</span>";
		}
		return value;
	},

	onload: function (report) {
		report.page.add_inner_button(__("Export to Excel"), function () {
			const data = frappe.query_report.data;
			if (!data || !data.length) {
				frappe.msgprint(__("No data to export."));
				return;
			}
			function fnum(val, dec) {
				var n = parseFloat(val) || 0;
				if (!n) return "";
				return n.toLocaleString("en-IN", { minimumFractionDigits: dec, maximumFractionDigits: dec });
			}
			function esc(s) {
				return (s || "").toString().replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;");
			}
			var sdate = frappe.query_report.get_filter_value("shift_date") || frappe.datetime.get_today();
			var sshift = frappe.query_report.get_filter_value("shift") || "Full Day";
			var heads = ["Time", "Voucher Type", "Voucher / GRN No", "Supplier / Department",
				"Item Code", "Item Description", "Warehouse", "Qty Received", "Qty Issued", "Balance In Stock", "UOM"];
			var html = '<html xmlns:o="urn:schemas-microsoft-com:office:office" '
				+ 'xmlns:x="urn:schemas-microsoft-com:office:excel" '
				+ 'xmlns="http://www.w3.org/TR/REC-html40"><head><meta charset="UTF-8">'
				+ '<!--[if gte mso 9]><xml><x:ExcelWorkbook><x:ExcelWorksheets><x:ExcelWorksheet>'
				+ '<x:Name>Store Shift Report</x:Name><x:WorksheetOptions><x:DisplayGridlines/>'
				+ '</x:WorksheetOptions></x:ExcelWorksheet></x:ExcelWorksheets></x:ExcelWorkbook></xml><![endif]-->'
				+ '</head><body><table border="0" style="font-family:Calibri;font-size:12pt;">'
				+ '<tr><td colspan="11" style="font-size:14pt;font-weight:bold;">STORES DAILY SHIFT REPORT</td></tr>'
				+ '<tr><td colspan="11">Date: ' + esc(sdate) + '&nbsp;&nbsp;&nbsp;Shift: ' + esc(sshift) + '</td></tr>'
				+ '<tr><td colspan="11"></td></tr></table>'
				+ '<table border="1" style="border-collapse:collapse;font-family:Calibri;font-size:11pt;">';
			html += '<tr style="background-color:#1a3c5e;color:#fff;font-weight:bold;">';
			heads.forEach(function (h, i) {
				html += '<td style="padding:6px 10px;' + (i >= 7 ? "text-align:right;" : "") + '">' + h + '</td>';
			});
			html += '</tr>';
			data.forEach(function (r) {
				if (r.row_type === "section") {
					html += '<tr style="background-color:#dce6f1;font-weight:bold;color:#1a3c5e;">'
						+ '<td colspan="11" style="padding:5px 8px;">' + esc(r.voucher_type) + '</td></tr>';
					return;
				}
				if (r.row_type === "subtotal") {
					html += '<tr style="background-color:#eaf2fb;font-weight:bold;">'
						+ '<td colspan="7" style="padding:4px 8px;text-align:right;">' + esc(r.party) + '</td>'
						+ '<td style="padding:4px 8px;text-align:right;">' + fnum(r.in_qty, 3) + '</td>'
						+ '<td style="padding:4px 8px;text-align:right;">' + fnum(r.out_qty, 3) + '</td>'
						+ '<td></td><td></td></tr>';
					return;
				}
				html += '<tr>'
					+ '<td style="padding:4px 8px;">' + esc(r.posting_time) + '</td>'
					+ '<td style="padding:4px 8px;">' + esc(r.voucher_type) + '</td>'
					+ '<td style="padding:4px 8px;">' + esc(r.voucher_no) + '</td>'
					+ '<td style="padding:4px 8px;">' + esc(r.party) + '</td>'
					+ '<td style="padding:4px 8px;">' + esc(r.item_code) + '</td>'
					+ '<td style="padding:4px 8px;">' + esc(r.item_name) + '</td>'
					+ '<td style="padding:4px 8px;">' + esc(r.warehouse) + '</td>'
					+ '<td style="padding:4px 8px;text-align:right;">' + fnum(r.in_qty, 3) + '</td>'
					+ '<td style="padding:4px 8px;text-align:right;">' + fnum(r.out_qty, 3) + '</td>'
					+ '<td style="padding:4px 8px;text-align:right;">' + fnum(r.balance_qty, 3) + '</td>'
					+ '<td style="padding:4px 8px;">' + esc(r.uom) + '</td></tr>';
			});
			html += '</table></body></html>';
			var blob = new Blob([html], { type: "application/vnd.ms-excel;charset=utf-8" });
			var url = URL.createObjectURL(blob);
			var a = document.createElement("a");
			a.href = url;
			a.download = "Store_Shift_Report_" + sdate + "_" + sshift + ".xls";
			document.body.appendChild(a);
			a.click();
			document.body.removeChild(a);
			URL.revokeObjectURL(url);
		});
	},
};

File: sandeep_reports/sandeep_reports/report/store_shift_report/store_shift_report.py
# Store Shift Report — standard Script Report (auto-converted from safe_exec DB script)
import frappe


def execute(filters=None):
	filters = frappe._dict(filters or {})
	# Store Shift Report — DB-stored Script Report (safe_exec / RestrictedPython compatible)
	# Contract: receives `filters`; must set `data` = (columns, rows, message, chart, report_summary)
	#
	# IMPORTANT safe_exec rule: functions defined here do NOT capture module-level
	# variables (no closures). Everything that touches the data is inlined.
	#
	# Shows store stock movements (Inward receipts + Outward issues) during a chosen
	# shift on a chosen date. Mirrors the "Stores Daily Shift Report" sheet so the
	# in-charge can pick a date + shift, then print or export to Excel.

	shift_date = filters.get("shift_date") or frappe.utils.nowdate()
	shift = filters.get("shift") or "Full Day"
	company = filters.get("company") or frappe.db.get_default("Company")
	movement = filters.get("movement") or "All"

	# ---- shift -> [start_dt, end_dt] window (Shift C crosses midnight into next day) ----
	next_day = frappe.utils.add_days(shift_date, 1)
	if shift == "A":
		start_dt = shift_date + " 06:00:00"
		end_dt = shift_date + " 13:59:59"
	elif shift == "B":
		start_dt = shift_date + " 14:00:00"
		end_dt = shift_date + " 21:59:59"
	elif shift == "C":
		start_dt = shift_date + " 22:00:00"
		end_dt = str(next_day) + " 05:59:59"
	else:
		start_dt = shift_date + " 00:00:00"
		end_dt = shift_date + " 23:59:59"

	# ---- filter conditions ----
	cond = ""
	params = {"co": company, "start": start_dt, "end": end_dt}

	# Warehouse (tree-inclusive: a parent/group warehouse includes all its children)
	if filters.get("warehouse"):
		wlr = frappe.db.get_value("Warehouse", filters.get("warehouse"), ["lft", "rgt"])
		if wlr and wlr[0] is not None:
			cond = cond + " AND sle.warehouse IN (SELECT w2.name FROM `tabWarehouse` w2 WHERE w2.lft >= %(wl)s AND w2.rgt <= %(wr)s)"
			params["wl"] = wlr[0]
			params["wr"] = wlr[1]
		else:
			cond = cond + " AND sle.warehouse = %(wh)s"
			params["wh"] = filters.get("warehouse")

	# Item Group (tree-inclusive)
	if filters.get("item_group"):
		glr = frappe.db.get_value("Item Group", filters.get("item_group"), ["lft", "rgt"])
		if glr and glr[0] is not None:
			cond = cond + " AND i.item_group IN (SELECT g2.name FROM `tabItem Group` g2 WHERE g2.lft >= %(igl)s AND g2.rgt <= %(igr)s)"
			params["igl"] = glr[0]
			params["igr"] = glr[1]
		else:
			cond = cond + " AND i.item_group = %(ig)s"
			params["ig"] = filters.get("item_group")

	# Item
	if filters.get("item_code"):
		cond = cond + " AND sle.item_code = %(item_code)s"
		params["item_code"] = filters.get("item_code")

	# Movement direction
	if movement == "Inward":
		cond = cond + " AND sle.actual_qty > 0"
	elif movement == "Outward":
		cond = cond + " AND sle.actual_qty < 0"

	query = (
		"SELECT sle.posting_date AS posting_date, sle.posting_time AS posting_time, "
		"sle.voucher_type AS voucher_type, sle.voucher_no AS voucher_no, "
		"sle.item_code AS item_code, i.item_name AS item_name, i.stock_uom AS uom, "
		"i.item_group AS item_group, sle.warehouse AS warehouse, "
		"sle.actual_qty AS actual_qty, sle.qty_after_transaction AS balance_qty "
		"FROM `tabStock Ledger Entry` sle "
		"INNER JOIN `tabItem` i ON i.name = sle.item_code "
		"WHERE sle.is_cancelled = 0 AND sle.company = %(co)s "
		"AND TIMESTAMP(sle.posting_date, sle.posting_time) >= %(start)s "
		"AND TIMESTAMP(sle.posting_date, sle.posting_time) <= %(end)s" + cond + " "
		"ORDER BY sle.posting_date, sle.posting_time, sle.voucher_no"
	)
	moves = frappe.db.sql(query, params, as_dict=True)

	# ---- party / source lookup per voucher (batched, no N+1, all inlined) ----
	by_type = {}
	for m in moves:
		by_type.setdefault(m["voucher_type"], set()).add(m["voucher_no"])

	party_map = {}
	specs = [
		["Purchase Receipt", "supplier_name"],
		["Purchase Invoice", "supplier_name"],
		["Delivery Note", "customer_name"],
		["Sales Invoice", "customer_name"],
		["Stock Entry", "stock_entry_type"],
	]
	for spec in specs:
		dt = spec[0]
		field = spec[1]
		names = by_type.get(dt)
		if names:
			for d in frappe.get_all(dt, filters={"name": ["in", list(names)]}, fields=["name", field]):
				party_map[dt + "::" + d["name"]] = d.get(field)

	# ---- split into inward / outward ----
	inward = []
	outward = []
	for m in moves:
		q = float(m.get("actual_qty") or 0)
		if q > 0:
			inward.append(m)
		elif q < 0:
			outward.append(m)

	rows = []
	tot_in = 0.0
	tot_out = 0.0

	if movement != "Outward":
		rows.append({"row_type": "section", "voucher_type": "MATERIAL RECEIPTS (INWARD)"})
		s = 0.0
		for m in inward:
			q = float(m.get("actual_qty") or 0)
			rows.append({
				"row_type": "data",
				"posting_time": m.get("posting_time"),
				"voucher_type": m.get("voucher_type"),
				"voucher_no": m.get("voucher_no"),
				"party": party_map.get(m["voucher_type"] + "::" + m["voucher_no"]) or "",
				"item_code": m.get("item_code"),
				"item_name": m.get("item_name"),
				"warehouse": m.get("warehouse"),
				"in_qty": q,
				"out_qty": 0.0,
				"balance_qty": float(m.get("balance_qty") or 0),
				"uom": m.get("uom"),
			})
			s = s + q
		tot_in = s
		rows.append({"row_type": "subtotal", "party": "Total Received", "in_qty": s, "out_qty": 0.0})

	if movement != "Inward":
		rows.append({"row_type": "section", "voucher_type": "MATERIAL ISSUES (OUTWARD)"})
		s = 0.0
		for m in outward:
			q = -float(m.get("actual_qty") or 0)
			rows.append({
				"row_type": "data",
				"posting_time": m.get("posting_time"),
				"voucher_type": m.get("voucher_type"),
				"voucher_no": m.get("voucher_no"),
				"party": party_map.get(m["voucher_type"] + "::" + m["voucher_no"]) or "",
				"item_code": m.get("item_code"),
				"item_name": m.get("item_name"),
				"warehouse": m.get("warehouse"),
				"in_qty": 0.0,
				"out_qty": q,
				"balance_qty": float(m.get("balance_qty") or 0),
				"uom": m.get("uom"),
			})
			s = s + q
		tot_out = s
		rows.append({"row_type": "subtotal", "party": "Total Issued", "in_qty": 0.0, "out_qty": s})

	# ---- summary cards ----
	report_summary = [
		{"value": len(inward), "label": "Receipt Lines", "datatype": "Int", "indicator": "blue"},
		{"value": tot_in, "label": "Qty Received", "datatype": "Float", "indicator": "blue"},
		{"value": len(outward), "label": "Issue Lines", "datatype": "Int", "indicator": "orange"},
		{"value": tot_out, "label": "Qty Issued", "datatype": "Float", "indicator": "orange"},
	]

	columns = [
		{"fieldname": "row_type", "label": "RT", "fieldtype": "Data", "hidden": 1},
		{"fieldname": "posting_time", "label": "Time", "fieldtype": "Data", "width": 80},
		{"fieldname": "voucher_type", "label": "Voucher Type", "fieldtype": "Data", "width": 150},
		{"fieldname": "voucher_no", "label": "Voucher / GRN No", "fieldtype": "Data", "width": 160},
		{"fieldname": "party", "label": "Supplier / Department", "fieldtype": "Data", "width": 180},
		{"fieldname": "item_code", "label": "Item Code", "fieldtype": "Link", "options": "Item", "width": 130},
		{"fieldname": "item_name", "label": "Item Description", "fieldtype": "Data", "width": 230},
		{"fieldname": "warehouse", "label": "Warehouse", "fieldtype": "Link", "options": "Warehouse", "width": 150},
		{"fieldname": "in_qty", "label": "Qty Received", "fieldtype": "Float", "precision": 3, "width": 120},
		{"fieldname": "out_qty", "label": "Qty Issued", "fieldtype": "Float", "precision": 3, "width": 120},
		{"fieldname": "balance_qty", "label": "Balance In Stock", "fieldtype": "Float", "precision": 3, "width": 130},
		{"fieldname": "uom", "label": "UOM", "fieldtype": "Data", "width": 70},
	]

	data = (columns, rows, None, None, report_summary)
	return data
